### PR TITLE
Deploy new resolver

### PR DESCRIPTION
**Pull request by DeRain**

### CODE REVIEW

File: contracts/Resolver.sol [hunk 1]
```diff
@@ -66,6 +69,56 @@ contract Resolver is SignatureUtil {
...
+    }
+
+    /**
+     * @dev Function get value by provied key hash. Keys hashes can be found in Sync event emitted by Registry.sol contract.
+     * @param keyHash The key to query the value of.
+     * @param tokenId The token id to set.
+     * @return The value string.
+     */
+    function getByHash(uint256 keyHash, uint256 tokenId) public view whenResolver(tokenId)  returns (string memory) {
+        return get(hashToKey(keyHash), tokenId);
```
Comment (bogdan) [inline contracts/Resolver.sol:103]: Return key and value

File: contracts/Resolver.sol [hunk 2]
```diff
@@ -66,6 +69,56 @@ contract Resolver is SignatureUtil {
...
+     * @return The values.
+     */
+    function getManyByHash(uint256[] memory keyHashes, uint256 tokenId) public view whenResolver(tokenId) returns (string[] memory) {
+        uint256 keyCount = keyHashes.length;
+        string[] memory values = new string[](keyCount);
+        uint256 preset = _tokenPresets[tokenId];
+        for (uint256 i = 0; i < keyCount; i++) {
+            values[i] = _records[tokenId][preset][hashToKey(keyHashes[i])];
+        }
+        return values;
```
Comment (bogdan) [inline contracts/Resolver.sol:119]: Return keys and values.

File: contracts/Resolver.sol [hunk 3]
```diff
@@ -159,9 +241,17 @@ contract Resolver is SignatureUtil {
      * @param tokenId uint256 ID of the token
      */
     function _set(uint256 preset, string memory key, string memory value, uint256 tokenId) internal {
-        _registry.sync(tokenId, uint256(keccak256(bytes(key))));
+        uint256 keyHash = uint256(keccak256(bytes(key)));
+        bool isNewKey = false;
+        _registry.sync(tokenId, keyHash);
+        if (bytes(_records[tokenId][preset][key]).length == 0) {
```
Comment (bogdan) [inline contracts/Resolver.sol:247]: `bool isNewKey = bytes(_records[tokenId][preset][key]).length == 0);`

File: GAS_ESTIMATIONS.md [hunk 4]
```diff
@@ -1,10 +1,30 @@
...
-      ⓘ Resolver.set: 60052 gas (~$0.0168, 140 USD/ETH)
-      ⓘ Resolver.setMany - one value: 62260 gas (~$0.0174, 140 USD/ETH)
-      ⓘ Resolver.setMany - two values: 94826 gas (~$0.0266, 140 USD/ETH)
-      ⓘ Resolver.setMany - three values: 127327 gas (~$0.0357, 140 USD/ETH)
-      ⓘ Resolver.setPreset: 49743 gas (~$0.0139, 140 USD/ETH)
-      ⓘ Resolver.reset: 34422 gas (~$0.0096, 140 USD/ETH)
+      ⓘ Resolver.set: 82742 gas (~$0.0232, 140 USD/ETH)
+      ⓘ Resolver.setMany - one value: 84972 gas (~$0.0238, 140 USD/ETH)
+      ⓘ Resolver.setMany - two values: 140241 gas (~$0.0393, 140 USD/ETH)
+      ⓘ Resolver.setMany - three values: 195446 gas (~$0.0547, 140 USD/ETH)
```
Comment (bogdan) [inline GAS_ESTIMATIONS.md:8]: Add tests to `setManyFor` for existing keys.

File: contracts/Resolver.sol [hunk 5]
```diff
@@ -8,15 +8,18 @@ import './controllers/MintingController.sol';
 
 contract Resolver is SignatureUtil {
 
-    event Set(uint256 indexed preset, string indexed key, string value, uint256 indexed tokenId);
-    event SetPreset(uint256 indexed preset, uint256 indexed tokenId);
+    event Set(uint256 indexed tokenId, uint256 indexed preset, bool indexed isNewKey, string key, string value);
```
Comment (mvlabat) [inline contracts/Resolver.sol:11]: Are `preset` and `isNewKey` values more important for us than `key` to make them indexed?

File: contracts/Resolver.sol [hunk 6]
```diff
@@ -145,23 +201,56 @@ contract Resolver is SignatureUtil {
...
+     * @dev Internal function to to set record. As opposed to set, this imposes no restrictions on msg.sender.
      * @param preset preset to set key/values on
      * @param key key of record to be set
      * @param value value of record to be set
      * @param tokenId uint256 ID of the token
      */
     function _set(uint256 preset, string memory key, string memory value, uint256 tokenId) internal {
-        _registry.sync(tokenId, uint256(keccak256(bytes(key))));
+        uint256 keyHash = uint256(keccak256(bytes(key)));
+        bool isNewKey = bytes(_records[tokenId][preset][key]).length == 0;
```
Comment (mvlabat) [inline contracts/Resolver.sol:247]: A micro-optimization is maybe possible here if we store `preset` mapping reference into a variable and then write to it without traversing the `_records` again. I'm not sure whether solidity optimizer handles that itself

File: migrations/2_deploy-contracts.js [hunk 7]
```diff
@@ -44,7 +44,7 @@ module.exports = (deployer, network, accounts) => {
       await sunriseController.renounceMinter()
     }
 
-    await deployer.deploy(Resolver, registry.address)
+    await deployer.deploy(Resolver, registry.address, sunriseController.address)
```
Comment (mvlabat) [inline migrations/2_deploy-contracts.js:47]: I wonder how it worked before without specifying the controller address O_o
Was it outdated?

File: test/Resolver.ganache.js [hunk 8]
```diff
@@ -74,5 +83,112 @@ contract('Resolver', function([coinbase, ...accounts]) {
...
+    
+    assert.isTrue(event.args.isNewKey)
+    
+    tx = await resolver.set('new-key', 'new-value', tok)
+    event = tx.logs.reduce(e => e.event == 'Set')
+
+    assert.isFalse(event.args.isNewKey)
+  })
+
+  it('should product correct Set event', async () => {
```
Comment (mvlabat) [inline test/Resolver.ganache.js:172]: nitpick: "should emit"

File: test/helpers/submitSigTransaction.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+          value: nonce,
+        },
+      ),
+      fromAccount,
+    )
+
+    return contractToSend[method + 'For'](...args, signature)
+  }
+
+module.exports = submitSigTransaction
```
Comment (mvlabat) [inline test/helpers/submitSigTransaction.js:31]: nitpick: newline :)

File: contracts/Resolver.sol (hunk 5 above)
Comment (DeRain) [reply contracts/Resolver.sol:11]: @mvlabat key actually can't be indexed - it's a dynamic string. The index can be bytes32 max. I believe that was an error in the old version of the contract.

File: migrations/2_deploy-contracts.js (hunk 7 above)
Comment (DeRain) [reply migrations/2_deploy-contracts.js:47]: @mvlabat not sure, but didn't work for me without the address at all.

File: contracts/Resolver.sol (hunk 5 above)
Comment (mvlabat) [reply contracts/Resolver.sol:11]: Your statement is not quite correct. Even dynamic types like strings and arrays can be indexed, the only difference is that their keccak256 hash will be stored in the topic instead

Comment (DeRain) [reply contracts/Resolver.sol:11]: @mvlabat not it won't be hashed according to low-level log API documentation, but just striped to bytes32 type.
https://solidity.readthedocs.io/en/v0.5.12/contracts.html#low-level-interface-to-logs

Comment (DeRain) [reply contracts/Resolver.sol:11]: @mvlabat you right, keccak256 will be stored instead according to this https://solidity.readthedocs.io/en/v0.5.12/contracts.html#events

**Pull request merged**